### DISCUSSION TITLE
[AI deck] Getting NanoFlowNet Obstacle Avoidance working

### DISCUSSION THREAD

**Discussion by boomer319** [bitcraze/discussions#1074] [Q&A]
Hello,

I am trying to get the [NanoFlowNet obstacle avoidance](https://github.com/gemenerik/gap8-obstacle-avoidance) working on the AI deck (as seen in the [blogpost](https://www.bitcraze.io/tag/ai-deck/)).

Unfortunately I am running into multiple errors while `make all` (I have seen the [AI deck workshop 1](https://youtu.be/o9asYPHxEB4?t=2542) and read through the [Makefile](https://github.com/gemenerik/gap8-obstacle-avoidance/blob/main/Makefile). So I assume `make all` is what needs to be done when building packages for the AI deck).

The goal here is to quantize the neural network in order to reduce it's size and make it be able to be executed on the AI deck alongside the code needed for the obstacle avoidance.

Error number 1 is likely caused when or before quantizing:

1. `ArithmeticError: Expression QUANTIZE_0_455 evaluated to incorrect dtype`

Error number 2 is just a reaffirmation of this thought. No quantization has happened.

2. `The opened graph must be quantized to use this command. Run the aquant command.`

The Make process continues despite the first two errors. The terminal output stops with the following line:

3. `ModuleNotFoundError: No module named 'stats'`

The [stats](https://github.com/GreenWaves-Technologies/gap_sdk/tree/master/tools/nntool/nntool/stats) module referred to here is a custom python module used by a GapSDK tool called nntool.

Error 3 can be avoided by exporting the python path to the respective module directory.

`export PYTHONPATH=$PYTHONPATH: /path/to/gap_sdk/tools/nntool/nntool`

Remember: The Makefile continues despite the errors.
Because there likely was no quantization being done, another error pops up. But hope dies last I guess.

4. `nntool.quantization.range_provider.RangesMissingError: no ranges found for DEPTHWISE_CONV_2D_0_128`

The errors might not seem to be directly related to anything upon first glance. But: The neural network .tflite file can be viewed in an online tool called [netron](https://netron.app/).

The node Quantize 455...as in `QUANTIZE_0_455` in Error 1):
![Screenshot from 2023-12-06 16-10-33](https://github.com/bitcraze/discussions/assets/129183102/7adf13ba-0730-432c-b56e-ba222f31f84e)

The node DepthwiseConv2D 128...as in `DEPTHWISE_CONV_2D_0_128`:
![Screenshot from 2023-12-06 16-10-05](https://github.com/bitcraze/discussions/assets/129183102/4cc1ced8-186b-4288-9dca-3297c2b1ad43)

I have already opened an issue in the [respective git repo](https://github.com/gemenerik/gap8-obstacle-avoidance/tree/main). I was hoping to broaden the spectrum of responses by opening this discussion.

Things I made sure to be doing:

- I have sourced the [GapSDK](https://github.com/GreenWaves-Technologies/gap_sdk) config for the AI deck: ai_deck.sh. Most recently I tried sourcing the GAP8_V3.sh as this is mentionend in the [Makefile](https://github.com/gemenerik/gap8-obstacle-avoidance/blob/main/Makefile).
- I have exported the correct JTAG config for my adapter: `export GAPY_OPENOCD_CABLE=~/verrueckter_schwarm/gap_sdk/utils/gap8-openocd/tcl/interface/ftdi/olimex-arm-usb-tiny-h.cfg`
- I have tried running `make clean all run platform=board
` as in [AI deck workshop 1](https://youtu.be/o9asYPHxEB4?t=2542)...surprise: another error.

5. `FileNotFoundError: [Errno 2] No such file or directory: 'BUILD_MODEL/nanoflownet_unquantized.json'`

`BUILD_MODEL` is a directory that seems to be created when running `make <...>`. This error is fixed by moving nanoflownet_unquantized.json into `BUILD_MODEL` after cleaning. After this you can only run `make` without the clean, reproducing the earlier mentioned errors 1 - 4.

Best regards

**Comment by knmcguire** [comment]
Hi! The aideck workshop is already on the older side, so there are already some alternative approaches you can or need to use. If you already went through the [getting started tutorial ](https://www.bitcraze.io/documentation/tutorials/getting-started-with-aideck/)of the aideck, you can look at the [classification example](https://www.bitcraze.io/documentation/repository/aideck-gap8-examples/master/ai-examples/classification-demo/). That is probably the approach you'll need to take.

About this repository itself, this is not our work, unfortunately (guestblogpost) so internally at Bitcraze we can't help much. @gemenerik would be the best to help out here, and I see that you have started some issues at the nanoflow repo. If there is no reply on this, perhaps try to contact the authors by email instead.

**Reply by boomer319** [reply]
I did try the classification example. It didn't work as expected (https://github.com/orgs/bitcraze/discussions/1018#discussion-5806895). But I think you are referring to the steps involved. I'll look into that.

As far as I'm concerned, @gemenerik was an intern of yours? NanoFlowNet Obstacle Avoidance was the result of his work? I understand that people can be busy. That's why I was hoping someone worked with him in a team or there was someone supervising him at bitcraze.

I found an Email address in the Paper related to the NanoFlowNet Obstacle Avoidance project. It isn't active anymore. Contacting the person through another email found through a quick google search lead to no response.

Regardless, it would be sad to see such an amazing project go unnoticed.

I want to get this to work.

Thanks for the reply!

**Reply by gemenerik** [reply]
Hi boomer319. I have seen your issues on the repository. I am maintaining the project in my spare time so it may take longer for me to reply. My work on NanoFlowNet was not part of my internship at Bitcraze.

**Reply by boomer319** [reply]
@gemenerik I figured you might have a lot going on. That's why I wanted to make a nice descriptive readme for the project. So you don't have to explain what to do to people as time goes on.

As far as bitcraze is concerned: This project shows the capabilites of the AI deck as much as, if not more, than the classification demo. Integrating this as another official example could be beneficial to show how the AI deck enhances the crazyflie. That's why I was drawn to this project in the first place.
